### PR TITLE
Improved AppConfig logic for 1.7.x and 1.8.x (EXPOSUREAPP-3455)

### PR DESCRIPTION
**Pull request by d4rken**
This PR refactores the `AppConfigProvider` to prepare for the changes to the app config in 1.7.x and 1.8.x

* The app config (`ConfigData`) consists now of "per module" sub interfaces, e.g. there is a `RiskCalculationConfig` that is created by the `RiskCalculationConfig.Mapper` such that changes to individual parts of the original protobuf config are hidden and can be worked on in parallel.
* The `AppConfigProvider` now offers a `val appConfig: Flow<ConfigData>` to allow other app components to react to config changes (e.g. reset RiskCalculuation, EXPOSUREAPP-2718 @chris-cwa )
* `ConfigData` exposes time variables (server time and the offset to our local time) to allow other app components to abort e.g. Riskcalclulation if the time is incorrect (EXPOSUREAPP-2998)
* There is a new test menu solely for viewing the current app config, seeing when it was last refreshed and to force a refresh.

## Test
* Smoke test the app
* Review that the tests in the `appconfig` package cover all edge cases you can think off
* Play a bit with the app config test menu

### CODE REVIEW

**Comment by d4rken** [comment]
Can be reviewed, still fixing the last unit tests.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/download/AppConfigServer.kt [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                }
+            }
+            if (exportBinary == null || exportSignature == null) {
+                throw ApplicationConfigurationInvalidException()
+            }
+
+            if (verificationKeys.hasInvalidSignature(exportBinary, exportSignature)) {
+                throw ApplicationConfigurationCorruptException()
+            }
+            exportBinary!!
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/download/AppConfigServer.kt:60]: !! shouldn't be used, change it so that the compiler regards it as non-null

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/CWAConfigMapper.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun AppConfig.ApplicationConfiguration.getMappedSupportedCountries(): List<String> {
+        if (supportedCountriesList == null) return emptyList()
+
+        if (supportedCountriesList.size == 1 && !VALID_CC.matches(supportedCountriesList.single())) {
+            Timber.w("Invalid country data, clearing. (%s)", supportedCountriesList)
+            return emptyList()
+        }
+
+        return supportedCountriesList
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/CWAConfigMapper.kt:31]: convert to when?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/ConfigData.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import de.rki.coronawarnapp.appconfig.mapping.ConfigMapping
+import org.joda.time.Duration
+import org.joda.time.Instant
+
+interface ConfigData : ConfigMapping {
+
+    val updatedAt: Instant
+
+    /**
+     * serverTime + localOffset = updatedAt
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/ConfigData.kt:12]: minor: shouldn't this comment rather go to updatedAt

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/DefaultConfigData.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import org.joda.time.Duration
+import org.joda.time.Instant
+
+data class DefaultConfigData(
+    val serverTime: Instant,
+    val mappedConfig: ConfigMapping,
+    override val localOffset: Duration,
+    override val isFallback: Boolean
+) : ConfigData, ConfigMapping by mappedConfig {
+    override val updatedAt: Instant = serverTime.plus(localOffset)
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/DefaultConfigData.kt:13]: what about 
"* If **[isFallback]** returns true,
     * you should probably ignore the time offset."

**Review by ralfgehrer — COMMENTED** [review]
Great work.

**Review by ralfgehrer — DISMISSED** [review]

**Comment by d4rken** [comment]
Offline mode behavior is not working correctly yet...

**Comment by d4rken** [comment]
Make sure to check offline (airplane mode) behavior too.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigProvider.kt [hunk 5]
```diff
@@ -1,149 +1,45 @@
...
-            }
-        }
-
-        val countryCheck = sanityChecked.supportedCountriesList
-        if (countryCheck.size == 1 && !VALID_CC.matches(countryCheck.single())) {
-            Timber.w("Invalid country data, clearing. (%s)", this.supportedCountriesList)
-            sanityChecked = sanityChecked.toNewConfig {
-                clearSupportedCountries()
+    suspend fun getAppConfig(): ConfigData = mutex.withLock {
+        Timber.tag(TAG).v("getAppConfig()")
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigProvider.kt:34]: Could be removed IMHO

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigSource.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    private val parser: ConfigParser,
+    private val dispatcherProvider: DispatcherProvider
+) {
+
+    suspend fun retrieveConfig(): ConfigData = withContext(dispatcherProvider.IO) {
+        Timber.v("retrieveConfig()")
+        val (serverBytes, serverError) = try {
+            server.downloadAppConfig() to null
+        } catch (e: Exception) {
+            Timber.tag(TAG).w(e, "Failed to download AppConfig from server .")
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigSource.kt:26]: rather an error (.e.) log?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/download/AppConfigServer.kt [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            )
+            Instant.parse(rawDate, DATE_FORMAT)
+        } catch (e: Exception) {
+            Timber.e("Failed to get server time.")
+            localTime
+        }
+        val offset = Duration(serverTime, localTime)
+        Timber.tag(TAG).v("Time offset was %dms", offset.millis)
+        return ConfigDownload(
+            rawData = rawConfig,
```
Comment (ralfgehrer) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/download/AppConfigServer.kt:75]: above "calculations" could be split into dedicated functions for easier readability

**Review by ralfgehrer — DISMISSED** [review]
Successfully tested with and without internet connection on a Pixel (Android 10).

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigProvider.kt (hunk 5 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigProvider.kt:34]: I like a bit more verbose logging :blush:

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigSource.kt (hunk 6 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigSource.kt:26]: As this would also trigger when no internet is available, it's more of a warning, isn't it? An error will be thrown and logged if then also no fallback config is available.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/download/AppConfigServer.kt (hunk 7 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/download/AppConfigServer.kt:63]: Moved the extraction into local extension functions to improve readability.

**Review by ralfgehrer — DISMISSED** [review]

**Review by harambasicluka — COMMENTED** [review]
tested on device, works like expected 👍

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+class ByteArrayAdapter : TypeAdapter<ByteArray>() {
+    override fun write(out: JsonWriter, value: ByteArray?) {
+        if (value == null) out.nullValue()
+        else value.toByteString().base64().let { out.value(it) }
+    }
+
+    override fun read(reader: JsonReader): ByteArray? = when (reader.peek()) {
+        NULL -> reader.nextNull().let { null }
+        else -> reader.nextString().decodeBase64()!!.toByteArray()
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt:18]:
```suggestion
        else -> reader.nextString().decodeBase64()?.toByteArray() ?: byteArrayOf()
```

**Review by chris-cwa — CHANGES_REQUESTED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt (hunk 8 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt:18]: alternativ:
```suggestion
        else -> reader.nextString().decodeBase64()?.toByteArray()
```

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt:18]: This would change the returned data though, if we are not reading a NULL token, then we expect there to be valid data, we want to crash then, not return "null" as if there would be no data.

AFAICT (see [here](https://github.com/square/okio/blob/master/okio/src/commonMain/kotlin/okio/-Base64.kt#L31)), the only way for `decodeBase64` to return null is on invalid data. So if we then return null in the not `NULL` token case, then we treat not existing data and invalid data the same. Would that be desired behavior?

Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt:18]: ok, if a crash is ok with everyone - let's do it!

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt:18]: I think the crash needs to happen, otherwise the classes above, using gson (with this adapter) to deserialize something, wouldn't know that there is something wrong with the data if there was no exception. What do you think?

Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt:18]: I believe this app should never crash in production. :( but my opinion is not the only one that counts.

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt:18]: That's not what I meant.

The code we are talking about is part of a custom type adapter for data serialization via Gson.

The TypeAdapter does not have enough information to determine whether returning `null` or throwing an exception is the correct approach, so it should not make that decision, but throw an exception so that the "higher level classes" know about this issue and can make the decision as they have more information available.

```kotlin
    fun getLastTimeStampData() : Instant? {
        return try {
            gson.fromJson(someFile)
        } catch (e: Exception) {
            Timber.e(e, "Our data was corrupted, returning EPOCH")
            Instant.EPOCH
        }
    }
```

The `getLastTimeStampData` knows about the data's purpose and can make a decision whether we should return null in an error case, or a fallback value. If we catch the decoding issue in the `ByteArrayAdapter` and just return `null`, then `getLastTimeStampData` could no longer distinguish between "no value exists" and "value exists but data corrupted/invalid".

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt:18]: Thinking about it, it would be good to throw a more specific Exception than a Kotlin NPE. Will add that.

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt]: Let me add a unit test for before it's merged.

**Review by chris-cwa — DISMISSED** [review]

**Review by chris-cwa — APPROVED** [review]

**Review by ralfgehrer — APPROVED** [review]

**Review by kolyaopahle — APPROVED** [review]

**Review by AlexanderAlferov — APPROVED** [review]
